cmd/memtierd/main.go

```diff
@@ -0,0 +1,339 @@
+	log.Infof("Connected to %s/%s...", runtime, version)
+
+	if config == "" {
+		return 0, nil
+	}
+
+	return 0, nil
+}
+
+func (p *plugin) StartContainer(ctx context.Context, pod *api.PodSandbox, ctr *api.Container) error {
+	log.Infof("Starting container %s/%s/%s...", pod.GetNamespace(), pod.GetName(), ctr.GetName())
+
+	hostRoot := opt.HostRoot
+
+	podName := pod.GetName()
+	containerName := ctr.GetName()
+	annotations := pod.GetAnnotations()
+
+	// If memtierd annotation is not present, don't execute further
+	class, ok := annotations["class.memtierd.nri"]
+	if !ok {
+		return nil
+	}
+
+	// Check that class is of correct form
+	pattern := "^[A-Za-z0-9_-]+$"
+	regex, err := regexp.Compile(pattern)
+	if err != nil {
+		// Handle error if the pattern is invalid
+		log.Fatalf("Invalid regex pattern:", err)
+		return nil
+	}
+
+	if !regex.MatchString(class) {
+		log.Fatalf("Invalid memtierd.class.nri!")
+		return nil
+	}
+
+	// You can specify the template here based on the given class ex. class.memtierd.nri: low-prio -> template = low-prio.yaml
+	// Plugin looks in the /template directory and looks for the low-prio.yaml then
+	template := ""
+
+	if class == "example-configuration" {
+		template = "example-configuration.yaml"
+	}
+
+	fullCgroupPath := getFullCgroupPath(ctr)
+	podDirectory, outputFilePath, configFilePath := editMemtierdConfig(fullCgroupPath, podName, containerName, template, hostRoot)
+	startMemtierd(podName, containerName, podDirectory, outputFilePath, configFilePath)
+
+	return nil
+}
+
+func (p *plugin) StopContainer(ctx context.Context, pod *api.PodSandbox, ctr *api.Container) ([]*api.ContainerUpdate, error) {
+	log.Infof("Stopped container %s/%s/%s...", pod.GetNamespace(), pod.GetName(), ctr.GetName())
+
+	//
+	// This is the container (post-)stop request handler. You can update any
+	// of the remaining running containers. Take a look at the functions in
+	// pkg/api/update.go to see the available controls.
+	//
+
+	podName := pod.GetName()
+	dirPath := fmt.Sprintf("%s/memtierd/%s", os.TempDir(), podName)
+
+	// Kill the memtierd process
+	out, err := exec.Command("sudo", "pkill", "-f", dirPath).CombinedOutput()
+	if err != nil {
+		exitErr, ok := err.(*exec.ExitError)
+		if !ok || exitErr.ExitCode() != 1 {
+			// Error occurred that is not related to "no processes found"
+			log.Fatalf("Error killing memtierd process: %v. Output: %s\n", err, out)
+		} else {
+			// "No processes found" error, do nothing
+			log.Printf("No processes found for memtierd process\n")
+		}
+	}
+
+	err = os.RemoveAll(dirPath)
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	return []*api.ContainerUpdate{}, nil
+}
+
+func (p *plugin) onClose() {
+	log.Infof("Connection to the runtime lost, exiting...")
+	os.Exit(0)
+}
+
+func getFullCgroupPath(ctr *api.Container) string {
+	cgroupPath := ctr.Linux.CgroupsPath
+
+	split := strings.Split(cgroupPath, ":")
+
+	partOne := split[0]
+	partTwo := fmt.Sprintf("%s-%s.scope", split[1], split[2])
```

Comment:
What happens is kubelet/the runtime is configured to run with the cgroupfs cgroup driver instead of systemd ? Is that caught and reported here or somewhere else ?